src/memory.rs

```diff
@@ -0,0 +1,124 @@
+use bootloader::bootinfo::*;
+use x86_64::{
+    PhysAddr, VirtAddr,
+    registers::control::Cr3,
+    structures::paging::{
+        FrameAllocator, Mapper, OffsetPageTable, Page, PageTable, PageTableFlags, PhysFrame,
+        Size4KiB,
+    },
+};
+
+/// # Safety
+/// This is unsafe because the caller must guarantee that memory at
+/// passed physical memory offset is already mapped to virtual memory.
+/// The caller must also guarantee that this function is only called once
+/// as the function returns a mutable reference and having multiple mutable references
+/// would violate Rust's aliasing rules and cause undefined behavior.
+pub unsafe fn init(phys_mem_offset: VirtAddr) -> OffsetPageTable<'static> {
+    unsafe {
+        let l4table = get_active_level_4_table(phys_mem_offset);
+        OffsetPageTable::new(l4table, phys_mem_offset)
+    }
+}
+
+pub fn create_mapping(
+    page: Page,
+    mapper: &mut OffsetPageTable,
+    frame_allocator: &mut impl FrameAllocator<Size4KiB>,
+) {
+    let frame = PhysFrame::containing_address(PhysAddr::new(0xb8000));
+    let flags = PageTableFlags::PRESENT | PageTableFlags::WRITABLE;
+    let map_to_result = unsafe { mapper.map_to(page, frame, flags, frame_allocator) };
+    map_to_result.expect("map_to failed").flush();
+}
+
+pub struct BootInfoFrameAllocator {
+    memory_map: &'static MemoryMap,
+    next: usize,
+}
```

Comment:
Let's insert blank lines between items like functions, types etc. for cleaner style.
Unfortunately there's no stable rustfmt option that will do this automatically.
```suggestion
}

```